src/main/java/io/netty/buffer/api/ComponentProcessor.java

```diff
@@ -0,0 +1,169 @@
+         *
+         * @param index The current index of the given buffer component, based on the initial index passed to the
+         *              {@link Buf#forEachWritable(int, OfWritable)} method.
+         * @param component The current buffer component being processed.
+         * @return {@code true} if the iteration should continue and more components should be processed, otherwise
+         * {@code false} to stop the iteration early.
+         */
+        boolean process(int index, WritableComponent component);
+    }
+
+    /**
+     * A view onto the buffer component being processed in a given iteration of
+     * {@link Buf#forEachReadable(int, OfReadable)}.
+     */
+    interface ReadableComponent {
+
+        /**
+         * Check if this component is backed by a cached byte array than can be accessed cheaply.
+         * <p>
+         * <strong>Note</strong> that regardless of what this method returns, the array should not be used to modify the
+         * contents of this buffer component.
+         *
+         * @return {@code true} if {@link #readableArray()} is a cheap operation, otherwise {@code false}.
+         */
+        boolean hasReadableArray();
+
+        /**
+         * Get a byte array of the contents of this component.
+         * <p>
+         * <strong>Note</strong> that the array is meant to be read-only. It may either be a direct reference to the
+         * concrete array instance that is backing this component, or it is a fresh copy.
+         * Writing to the array may produce undefined behaviour.
+         *
+         * @return A byte array of the contents of this component.
+         * @throws UnsupportedOperationException if {@link #hasReadableArray()} returns {@code false}.
+         */
+        byte[] readableArray();
+
+        /**
+         * An offset into the {@link #readableArray()} where this component starts.
+         *
+         * @return An offset into {@link #readableArray()}.
+         * @throws UnsupportedOperationException if {@link #hasReadableArray()} returns {@code false}.
+         */
+        int readableArrayOffset();
+
+        /**
+         * Give the native memory address backing this buffer, or return 0 if this buffer has no native memory address.
+         * <p>
+         * <strong>Note</strong> that the address should not be used for writing to the buffer memory, and doing so may
+         * produce undefined behaviour.
+         *
+         * @return The native memory address, if any, otherwise 0.
+         */
+        long readableNativeAddress();
+
+        /**
+         * Get a {@link ByteBuffer} instance for this memory component.
+         * <p>
+         * <strong>Note</strong> that the {@link ByteBuffer} is read-only, to prevent write accesses to the memory,
+         * when the buffer component is obtained through {@link Buf#forEachReadable(int, OfReadable)}.
+         *
+         * @return A new {@link ByteBuffer} for this memory component.
```

Comment:
maybe just add something like `... which has its own position / limit`. I know this is kind of already hinted with `New` but maybe it makes it easier for people to understand the semantics.